lib/filterx/expr-null-coalesce.c

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2024 shifter <shifter@axoflow.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx/expr-null-coalesce.h"
+#include "filterx/object-null.h"
+#include "filterx/filterx-eval.h"
+
+struct _FilterXNullCoalesce
+{
+  FilterXBinaryOp super;
+};
+
+static FilterXObject *
+_eval(FilterXExpr *s)
+{
+  FilterXNullCoalesce *self = (FilterXNullCoalesce *) s;
+
+  FilterXObject *lhs_object = NULL;
+  FilterXObject *rhs_object = NULL;
+
+  if (self->super.lhs)
+    lhs_object = filterx_expr_eval(self->super.lhs);
+  if (!lhs_object || filterx_object_is_type(lhs_object, &FILTERX_TYPE_NAME(null)))
+    {
+      if (!lhs_object)
+        {
+          msg_debug("FILTERX NULL coalesce error handling",
+                    filterx_format_last_error());
```

Comment:
I think this could be a bit more verbose, to communicate the behavior cleaner, mentioning that the lhs evaluated with an error, so the rhs will be returned.